src/main/java/net/minecraftforge/common/world/RegistryAccessExtension.java

```diff
@@ -0,0 +1,123 @@
+import net.minecraftforge.registries.ForgeRegistryEntry;
+import net.minecraftforge.registries.IForgeRegistryEntry;
+import net.minecraftforge.registries.RegistryBuilder;
+
+/**
+ * Defines an additional world-gen component to load as part of the RegistryAccess registries.
+ * <p>
+ * The path of the provided ResourceKey becomes the data folder from which datapack definitions are parsed
+ * and the provided Codec handles encoding and decoding to and from the data format.
+ * <p>
+ * To avoid conflicts with other mods, the path of the ResourceKey must be prefixed with the owning mod's
+ * namespace (ie {@code <modid>:<modid>/path}). Static helper methods for creating valid ResourceKeys are
+ * provided at {@link #createRegistryKey(String, String)} and {@link #createRegistryKey(ResourceLocation)}.
+ * <p>
+ * Default or 'built-in' registry entries can be registered via a forge DeferredRegister or IForgeRegistry
+ * of the same registry ResourceKey. These elements will populate the resulting RegistryAccess before datapacks
+ * are loaded and may be overridden by data-definitions. A static helper method for creating a DeferredRegisters
+ * for a RegistryAccessExtension is provided at {@link #createRegister(Class, ResourceKey)}.
+ *
+ * @param <T> The type of world-gen component.
+ */
+public class RegistryAccessExtension<T extends IForgeRegistryEntry<T>> extends ForgeRegistryEntry<RegistryAccessExtension<?>>
+{
+    private final ResourceKey<Registry<T>> registryKey;
+    private final Codec<T> directCodec;
+    private final Lifecycle defaultElementLifecycle;
+
+    public RegistryAccessExtension(ResourceKey<Registry<T>> registryKey, Codec<T> directCodec)
+    {
+        this(registryKey, directCodec, Lifecycle.experimental());
+    }
+
+    public RegistryAccessExtension(ResourceKey<Registry<T>> registryKey, Codec<T> directCodec, Lifecycle defaultElementLifecycle)
+    {
+        validateRegistryKey(registryKey);
+        setRegistryName(registryKey.location());
+        this.registryKey = registryKey;
+        this.directCodec = directCodec;
+        this.defaultElementLifecycle = defaultElementLifecycle;
+    }
+
+    public ResourceKey<? extends Registry<T>> getRegistryKey()
+    {
+        return registryKey;
+    }
+
+    public Codec<T> getDirectCodec()
+    {
+        return directCodec;
+    }
+
+    public Lifecycle getDefaultElementLifecycle()
+    {
+        return defaultElementLifecycle;
+    }
+
+    public static <T> ResourceKey<Registry<T>> createRegistryKey(String namespace, String path)
+    {
+        var prefix = getPathPrefix(namespace);
+        if (!path.startsWith(prefix))
+        {
+            path = prefix + path;
+        }
+
+        return ResourceKey.createRegistryKey(new ResourceLocation(namespace, path));
+    }
+
+    public static <T> ResourceKey<Registry<T>> createRegistryKey(ResourceLocation location)
+    {
+        return createRegistryKey(location.getNamespace(), location.getPath());
+    }
```

Comment:
This could be better handled as a static constructor for this class or as inputs to the constructor itself. It's more strange that you need to call these methods within the constructor itself.

Reply:
You wanted this class to _not_ be statically initiallized (previous comments on the test class) however we do need static references to the registry ResourceKeys in constructing Codecs that can be referenced by other dynamic components.

I added these helper methods so that users can create static ResourceKeys with the appropriate namespacing, and then extensions can just be initialized with those keys during registration.

For clarity are you suggesting go back to having statically init'd RegAccExtns? Or that mods should take responsibilty for creating the ResourceKeys with the appropriate path rather than having helper methods?

Reply:
A static constructor is not the same as a static initialization. A static constructor is simple a method to create an object via a static method. It can still be used to register an object lazily. However, the point I was making is that the registry key should be made as part of the constructor instead of having as a method that would need to be passed in.

```java
static Test create(String name, String path) {
   // create the registry key
}

// or

Test(String name, String path) {
  this.registryKey = // registry value
}
```

instead of

```java
Test(RegistryKey)

// Somewhere to call
Test(methodToCreateKey(String name, String path))
```

Reply:
If we’re using this PR to add a custom registry to the RegistryAccess then we’re holding reference to that registry’s ResourceKey in a static final field somewhere so that we can:
a) retrieve our registry from the RegistryHolder and actually use the component instances it contains.
b) create codecs (also typically static final fields) that can be used to reference our component from other components (the Supplier<T> codecs).

If we’re holding reference to the registry’s ResourceKey already we may as well be constructing the RegAccExtn with it.

Having helper methods that ‘correctly’ construct our registry ResourceKey that we can hold on to in a static field  is more useful to us than it being constructed inside a lazily computed object that we can’t